### PR TITLE
Publish Marian/OpusTrainer configuration YAMLs and dataset statistics

### LINKED ISSUE #313: Publish original Marian configs
Depends #309 

The publication code only supports Marian final configuration (from its own output).
But it would be useful to view the hierarchy of YAML files directly from Weight & Biases.

The hiearchy of files is merged in a specific order, so we need to investigate:
- how to get a final unique YAML file
- can we isolate the parsing code & run it from publication code
- or simply upload the YAML files as they are on-disk

### LINKED ISSUE #529: Display corpus size in W&B
We should display things we look at often in W&B. Final merged corpus size after deduplication is something I look at periodically to understand how aggressive the cleaning is overall. We can also display corpus size after each cleaning stage as we discussed with @gregtatum which should probably be a part of the analysis job.

### PR DESCRIPTION
**Pull request by vrigal**
Closes #313
Closes #529

### CODE REVIEW

**Comment by vrigal** [comment]
I tested based on branch `demo-online-publication` (https://github.com/vrigal/firefox-translations-training/commit/8eda3ad7205ccc905d839f388e28592c13cce931), with fake configuration files and datasets. Results are visible as artifacts on this run: https://wandb.ai/teklia/720-online?nw=nwuserrigal.

I'm pushing a commit that will be removed, in order to trigger a training in CI.

**Comment by vrigal** [comment]
I rebased on top of `vrigal:unique-wandb-runs`, so it will be easier for you to rebase once we merge it (by dropping the first commit).

I was actually able to publish all the artifacts from an online training task:
* Taskcluster: https://firefox-ci-tc.services.mozilla.com/tasks/MwT-4_l5TIOOObkKiyW6pQ
* W&B run artifacts: https://wandb.ai/moz-translations/ru-en/runs/uqqbbf6r/artifacts

**Comment by eu9ene** [comment]
> I rebased on top of `vrigal:unique-wandb-runs`, so it will be easier for you to rebase once we merge it (by dropping the first commit).
> 
> I was actually able to publish all the artifacts from an online training task:
> 
>     * Taskcluster: https://firefox-ci-tc.services.mozilla.com/tasks/MwT-4_l5TIOOObkKiyW6pQ
> 
>     * W&B run artifacts: https://wandb.ai/moz-translations/ru-en/runs/uqqbbf6r/artifacts

It's great that we were able to extract this info! However, I looked at https://wandb.ai/moz-translations/ru-en/runs/uqqbbf6r/artifacts and I don't think it will be convenient to use it in this form. Let's discuss live what we can do about it.

**Comment by eu9ene** [comment]
To reiterate what we discussed with @La0:
- dataset sizes should be published on the main run dashboard along custom charts as a single number widgets (grafana style) for visibility
- Extra Marian and OpusTrainer configs should be published in the main run config under different keys for easier navigation. Let's come up with the high-level keys for the new hierarchy of the run config: 
  - `marian` - full Marian config which is already being published
  - `model` - "configs/model/${model_type}.yml" Marian argument 
  - `training` - "configs/training/${model_type}.${training_type}.yml" Marian argument
  - `opustrainer` - OpusTrainer config
  - `extra-args` - extra Marian arguments from the experiment config

**Comment by vrigal** [comment]
@eu9ene I implemented the new publication. I restored the [TRASHME commit](https://github.com/mozilla/firefox-translations-training/pull/720/commits/607ab8eb5c6e26ae156982a3893fe11d2b10650f) (must be removed before merging). The new results are visible on the link below:
https://wandb.ai/moz-translations/ru-en/groups/ci_FGBmVk5oQNSb7D9_7GaLSw/workspace?nw=nwuserbabadiemoz

File: tracking/translations_parser/parser.py [hunk 1]
```diff
@@ -36,6 +40,10 @@
 # Expected version of Marian for a clean parsing
 SUPPORTED_MARIAN_VERSIONS = [(1, 10), (1, 12)]
 
+MARIAN_ARGS_REGEX = re.compile(r"command line:[\n ]+[\w\/-]+\/marian +(.*)")
+# Last Marian command line argument (not being part of training extra arguments)
+LAST_MARIAN_DECLARED_ARGUMENT = "seed"
```
Comment (eu9ene) [inline tracking/translations_parser/parser.py:45]: this is fragile and will change after the refactoring (rewriting the train.sh in python)

File: tracking/translations_parser/parser.py [hunk 2]
```diff
@@ -215,6 +223,100 @@ def _join(seq):
...
+            logger.info(
+                "Extra configuration files can only be retrieved in Taskcluster context, skipping."
+            )
+            return extra_config
+
+        # Handle Marian model and training YAML configuration files
+        if len(args["c"]) != 2:
+            logger.warning("Invalid")
+        else:
+            model_path, training_path = args["c"]
```
Comment (eu9ene) [inline tracking/translations_parser/parser.py:282]: the order is not guaranteed here. We can identify which config is which based on their names

File: tracking/translations_parser/parser.py [hunk 3]
```diff
@@ -215,6 +223,100 @@ def _join(seq):
...
+            except Exception as e:
+                logger.warning(f"Impossible to parse Marian model config at {model_path}: {e}")
+            try:
+                with open(training_path, "r") as f:
+                    extra_config["training"] = yaml.safe_load(f.read())
+            except Exception as e:
+                logger.warning(f"Impossible to parse Marian model config at {model_path}: {e}")
+
+        # Handle OpusTrainer configuration
+        (model_path,) = args.get("model", ("./model.npz",))
```
Comment (eu9ene) [inline tracking/translations_parser/parser.py:295]: When publishing from GCP this path will not mean anything useful.

File: tracking/translations_parser/parser.py (hunk 1 above)
Comment (eu9ene) [reply tracking/translations_parser/parser.py:45]: So, I don't think it's possible to distinguish what args are additional from inside the script. We can just log all the marian args then. Having all of them also can be useful.

**Review by eu9ene — CHANGES_REQUESTED** [review]
Even though it does what we want now, the implementation is quite fragile and will not withstand a major refactoring in train.sh (see https://github.com/mozilla/firefox-translations-training/issues/599). We should take measures to make it more robust before merging.

File: tracking/translations_parser/parser.py (hunk 3 above)
Comment (vrigal) [reply tracking/translations_parser/parser.py:292]: This should be covered by the `TASK_ID` check, logging `Extra configuration files can only be retrieved in Taskcluster context, skipping.` and returning early.

**Comment by vrigal** [comment]
@eu9ene I published new results with your suggestions: https://wandb.ai/moz-translations/ru-en/groups/ci_IYdCmIuJSTW3Cw3_9TaHbA

I'm removing the commit used to trigger the jobs in Taskcluster.

For a reminder, here are the 2 other groups created during this development (may help you cleaning up things):
* https://wandb.ai/moz-translations/ru-en/groups/ci_FGBmVk5oQNSb7D9_7GaLSw/
* https://wandb.ai/moz-translations/ru-en/groups/ci_ZVWvAVL9Quq3hOiEkQxKdg/

**Comment by eu9ene** [comment]
I see two datasets tables for some reason: https://wandb.ai/moz-translations/ru-en/runs/teacher-1_IYdCm/workspace?nw=nwuserepavlov

**Comment by vrigal** [comment]
> I see two datasets tables for some reason: https://wandb.ai/moz-translations/ru-en/runs/teacher-1_IYdCm/workspace?nw=nwuserepavlov

This is certainly related to https://github.com/mozilla/firefox-translations-training/issues/716#issuecomment-2278790264. When I use a different workspace (i.e. [nwuserbabadiemoz](https://wandb.ai/moz-translations/ru-en/groups/ci_IYdCmIuJSTW3Cw3_9TaHbA/workspace?nw=nwuserbabadiemoz)) everything seems fine.

**Comment by eu9ene** [comment]
As discussed I suggest we wait for #792 to ensure everything works correctly and then merge.

**Comment by eu9ene** [comment]
@vrigal I triggered the CI and there are two "group_logs" there https://wandb.ai/moz-translations/ci?nw=nwuserepavlov

**Comment by vrigal** [comment]
> @vrigal I triggered the CI and there are two "group_logs" there https://wandb.ai/moz-translations/ci?nw=nwuserepavlov

Nice catch, thank you ! I identified the problem and opened #820.

**Comment by eu9ene** [comment]
@vrigal there's an issue with `training` arguments for `finetune-student` step https://wandb.ai/moz-translations/ci/runs/finetune-student_ZrWxf/overview

Screenshot 2024-08-30 at 6 01 40 PM

**Comment by vrigal** [comment]
> @vrigal there's an issue with `training` arguments for `finetune-student` step https://wandb.ai/moz-translations/ci/runs/finetune-student_ZrWxf/overview

Thank you ! The script was detecting the `training` and `model` configs based on the end of the filename (`*.train.yml`). I updated it to detect base on the start of the string (`configs/model`, `configs/training`), which is logic. I'm pushing to CI again.

**Comment by vrigal** [comment]
New results have been published there: https://wandb.ai/moz-translations/ci/groups/ci_bT86ivvqSPSTJT0T2d-k_A/workspace?nw=nwuserepavlov
I'm removing the parameter to trigger CI.

**Review by eu9ene — APPROVED** [review]

**Pull request merged**